File: java/datastructures/lib/src/test/java/datastructures/queue/QueueTest.java
package datastructures.queue;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class QueueTest {
  @Test
  @DisplayName("enqueue into a queue")
  void enqueueSingleNodeIntoQueue(){
    Queue<Integer> queue = new Queue<>();
    queue.enqueue(1);
    System.out.println(queue);
    assertTrue(true);
  }

  @Test
  @DisplayName("enqueue multiple values into a queue")
  void enqueueMultiNodesIntoQueue() {
    Queue<Integer> queue = new Queue<>();
    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);
    System.out.println(queue);
    assertTrue(true);
  }

  @Test
  @DisplayName("dequeue out of the queue the expected value")
  void dequeueValueFromQueue() {
    Queue<Integer> queue = new Queue<>();
    queue.enqueue(1);
    queue.enqueue(2);
    System.out.println("First in, First out value in the queue is " + queue.peek());
    assertEquals(1, queue.dequeue());
  }

  @Test
  @DisplayName("peek into a queue and see the expected value")
  void peekAtFrontValueInQueue() {
    Queue<Integer> queue = new Queue<>();
    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);
    System.out.println("The front of the queue is " + queue.peek());
    assertEquals(1, queue.peek());
  }

  @Test
  @DisplayName("successfully empty queue after multiple dequeues")
  void emptyQueueAfterMultiDequeues(){
    Queue<Integer> queue = new Queue<>();
    queue.enqueue(1);
    queue.enqueue(2);
    queue.dequeue();
    queue.dequeue();
    assertThrows(IllegalArgumentException.class, queue::peek);
    assertTrue(queue.isEmpty());
  }

  @Test
  @DisplayName("successfully instantiate an empty queue")
  void instantiateAnEmptyQueue(){
    Queue<Integer> queue = new Queue<>();
    System.out.println("created an empty queue: " + queue);
    assertTrue(queue.isEmpty());
  }

  @Test
  @DisplayName("calling deque or peek on empty queue raises exception")
  void raiseExceptionOnEmptyQueue(){
    Queue<Integer> queue = new Queue<>();
    assertThrows(IllegalArgumentException.class, queue::dequeue);
    assertThrows(IllegalArgumentException.class, queue::peek);
  }

}

File: javascript/code-challenges/stack-and-queue/src/queue.test.ts
import { Queue } from "./queue";

describe("Queue", () => {
  it("can enqueue", () => {
    const q = new Queue<string>();

    expect(() => q.enqueue("Frodo")).not.toThrow();
  });

  it("enqueues", () => {
    const queue = new Queue<string>();

    queue.enqueue("Frodo");
    queue.enqueue("Sam");

    expect(queue.size).toBe(2);
    expect(queue.peek).toBe("Frodo");
  });

  it("can successfully queue into a queue", () => {
    const queue = new Queue<number>();

    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);

    expect(queue).toBeDefined();
    expect(queue.size).toBe(3);

  });

  it("can successfully queue multiple values into a queue", () => {
    const queue = new Queue<number>();

    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);
    queue.enqueue(4);
    queue.enqueue(5);

    expect(queue).toBeDefined();
    expect(queue.size).toBe(5);

  });

  it("can successfully dequeue out of a queue the expected value", () => {
    const queue = new Queue<number>();

    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);

    expect(queue).toBeDefined();
    expect(queue.size).toBe(3);
    expect(queue.dequeue()).toBe(3);

  });

  it("can successfully peek into a queue", () => {
    const queue = new Queue<number>();

    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);

    expect(queue).toBeDefined();
    expect(queue.size).toBe(3);
    expect(queue.peek).toBe(1);

  });

  it("can successfully empty a queue after multiple dequeues", () => {
    const queue = new Queue<number>();

    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);

    queue.dequeue();
    queue.dequeue();
    queue.dequeue();

    expect(queue).toBeDefined();
    expect(queue.size).toBe(0);

  });

  it("can successfully empty a queue after multiple dequeues", () => {
    const queue = new Queue<number>();

    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);

    queue.dequeue();
    queue.dequeue();
    queue.dequeue();

    expect(queue).toBeDefined();
    expect(queue.size).toBe(0);

  });

  it("can successfully instantiate an empty queue", () => {
    const queue = new Queue<number>();

    expect(queue).toBeDefined();
    expect(queue.size).toBe(0);

  });

  it("throws an error when dequeue on an empty queue", () => {
    const queue = new Queue<number>();

    expect(queue).toBeDefined();
    expect(() => {queue.dequeue()}).toThrow();

  });

  it("throws an error when peek on an empty queue", () => {
    const queue = new Queue<number>();

    expect(queue).toBeDefined();
    expect(() => {queue.peek}).toThrow();

  });
});
